masschallenge/impact-api
# [AC-5111] - Add SSO Token Generation, Verification and Refresh to impact-api

**Changes Introduced in [AC-5111](https://masschallenge.atlassian.net/browse/AC-5111):**

- add drf-jwt as dependency
- register urls for the following views:
  - obtain a jwt
  - verify a jwt
  - refresh a jwt
- configure drf-jwt in settings:
  - reasonable expiration deltas
  - a secret key (the production one is a secret travis env var.
  - work with JWTs in cookies.
- cleanup impact.urls.py (for clarity of urls.py chosen structure).


**Test:**
- make build;make dev;make load-remote-db;make grant-permissions (with v1_clients);
- update to the following [Postman Collection](https://www.getpostman.com/collections/0a95773f88df764b15ee)
- open a sample V1 call in postman, e.g. "v1 Industry list get" . Try to access it as is, get a 403.
- open sso/obtain-token request.
 - send incorrect username and password. see a 400 error
 -  send it with the correct username and password. see a 200 status, and the token
   - see that now there's a cookie added to "Cookies", with the same token.
- now reopen "v1 Industry list get":
  - send it. it should still fail (401), because you're also sending a (bad) token from our existing mechanism.
  - Go to the request "Headers" section. uncheck "Authorization". resend the request. It now succeeds.
- now create an application, and set the client_id in the "oauth token" request, and send the request (the old mechanism).
  - reselect the checkbox for Authorization in "v1 Industry list get", and send the request. it is successful. unselect, send, still successfull (the new and old mechanisms are compatible).

- in settings.py, change the value of JWT_EXPIRATION_DELTA to be very short (a few seconds). get a new token. wait that timedelta. try to use it. see that it fails (401). try to use this token in the "verify" call. see a 400 error.
- create a new token. before time is up, refresh it. see that it succeeds, and you can use it beyond the time you set - but no more than the time specified JWT_REFRESH_EXPIRATION_DELTA (set it to be short as well. Try to refresh after that time has passed. fails).


- review urlpatterns naming choices, and suggest improvements if there are any.
- review and groom https://masschallenge.atlassian.net/browse/AC-5110 and https://masschallenge.atlassian.net/browse/AC-5109 (in that order)
